rl/approximate_dynamic_programming.py

```diff
@@ -143,11 +143,8 @@ def return_(s_r: Tuple[S, float], i=i) -> float:
             return r + γ * (v[i-1].evaluate([s]).item() if i > 0 else 0.)
 
         v.append(
-            FunctionApprox.converged(sgd(
-                approx0,
-                repeat([(s, res.expectation(return_))
-                        for s, res in step.items()])
-            ))
+            approx0.solve([(s, res.expectation(return_))
+                           for s, res in step.items()])
```

Comment:
This needed a lot of changes. But looks like it's all working now! Much nicer to user Iterable than Sequence. I haven't changed all the calling code to pass along Iterators instead of Lists. Will do that later, on a new branch.

Reply:
Awesome.

I should note that taking a `Sequence` as an argument instead of an `Iterable` _can_ make sense—it depends on how you use it inside the function. If you only need to iterate over the values, or you're always going to build some new data structure from it (like a NumPy array), taking an `Iterable` makes sense. But if you're going to take the `Iterable` and turn it into a list anyway—maybe to index into it, for example—it would be better to take a `Sequence` as an input so that you don't need to create a new list if the caller has one already.

It's actually a bit of a tricky design question for interfaces and, if we don't know exactly how implementations of the interface will use their inputs, it's probably better to go for the more general option (`Iterable`). But it's a good thing to keep in mind.